upgrade.ts

```diff
@@ -0,0 +1,127 @@
+  /**
+   * Allows Angular v2 Component to be used from AngularJS v1.
+   */
+  downgradeNg2Component(type: Type): Function;
+  /**
+   * Bootstrap a hybrid AngularJS v1 / Angular v2 application.
+   */
+  bootstrap(element: Element, modules?: any[], config?: angular.IAngularBootstrapConfig): UgradeAdapterRef;
+  /**
+   * Adds a provider to the top level environment of a hybrid AngularJS v1 / Angular v2 application.
+   */
+  addProvider(provider: Type | any[] | any): void;
+  /**
+   * Allows Angular v2 service to be accessible from AngularJS v1.
+   */
+  downgradeNg2Provider(token: any): Function;
+}
+
+export class NgMetadataUpgradeAdapter {
+
+  bootstrap: Function;
+
+  private _upgradeAdapter: upgradeAdapterInstance;
+
+  constructor( UpgradeAdapter: UpgradeAdapter ) {
+    /**
+     * Manage the @angular/upgrade singleton
+     */
+    this._upgradeAdapter = new UpgradeAdapter();
+    /**
+     * Used to bootstrap a hybrid Angular 1 and Angular 2 application,
+     * using the same signature as `bootstrap` from ng-metadata/platform-browser-dynamic
+     *
+     * E.g. `upgradeAdapter.bootstrap(AppComponent, providers)`
+     */
+    this.bootstrap = createBootstrapFn(this._upgradeAdapter.bootstrap.bind(this._upgradeAdapter));
+  }
+
+  /**
+   * Used to register an Angular 2 component as a directive on an Angular 1 module,
+   * where the directive name is automatically created from the selector.
+   *
+   * E.g. `.directive(...upgradeAdapter.downgradeNg2Component(Ng2Component))
+   */
+  downgradeNg2Component( component: Type ): [ string, Function ] {
+    const annotations = reflector.annotations( component );
+    const cmpAnnotation = annotations[ 0 ];
+    const directiveName = resolveDirectiveNameFromSelector( cmpAnnotation.selector );
+    return [ directiveName, this._upgradeAdapter.downgradeNg2Component( component ) ];
+  }
+
+  /**
+   * Used to register an Angular 2 component by including it in the directives array
+   * of an ng-metadata annotated Angular 1 component.
+   *
+   * E.g.
+   * ```
+   * @Component({
+   *  selector: 'foo',
+   *  directives: [upgradeAdapter.provideNg2Component(Ng2Component)],
+   * })
+   * ```
+   */
+  provideNg2Component( component: Type ): PrecompiledDirectiveFactory {
+    const downgradedComponent = this.downgradeNg2Component( component );
+    const directiveFactory: PrecompiledDirectiveFactory = downgradedComponent[ 1 ];
+    directiveFactory.__precompiledAs = downgradedComponent[ 0 ];
+    return directiveFactory;
+  }
+
+  /**
+   * Adds an Angular 2 provider to the hybrid application.
+   */
+  addProvider( provider: Type | any[] | any ): void {
+    return this._upgradeAdapter.addProvider( provider );
+  }
+
+  /**
+   * Downgrades an Angular 2 Provider so that it can be registered as an Angular 1
+   * factory.
+   *
+   * E.g. `.factory('ng2Service', upgradeAdapter.downgradeNg2Provider(Ng2Service))`
+   */
+  downgradeNg2Provider( token: any ): Function {
```

Comment:
just an Idea, but maybe we can be consistent with our deprecated `provide` function

so instead of `.factory('ng2Service', upgradeAdapter.downgradeNg2Provider(Ng2Service))`
we could do something like:

`.factory(...upgradeAdapter.downgradeNg2Provider( 'ng2Service', {useClass:Ng2Service} ))`

or even better without string:

```typescript
export const Ng2ServiceToken = new OpaqueToken('ng2Service');
ngModule.factory(...upgradeAdapter.downgradeNg2Provider( Ng2ServiceToken, {useClass:Ng2Service} ))
```